<a href="https://colab.research.google.com/github/cemsah/Coursera_Capstone/blob/master/Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Capstone Project for the IBM Data Science Professional Certificate

This notebook is created for the Capstone Project for the IBM Data Science Professional Certificate on Coursera. 

## Table of Contents

#### Part 1 - Dataframe of Neighborhood Postal Codes

#### Part 2 - Coordinates of Postal Codes

#### Part 3 - Cluster Neighborhoods and Examine Clusters


Before starting the analysis I import libraries that will be used in the analysis. 

In [0]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes 
# uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## Part One - Dataframe of Neighborhood Postal Codes

Scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the postal codes data.

In [0]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [0]:
df=pd.read_html(url)[0]

In [0]:
df.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [0]:
df.shape

(288, 3)

The data frame has 288 rows of data. 

Before cleaning the data, I change the column names to match the assignment page. 

In [0]:
df.rename(columns={'Postcode':'PostalCode', 'Borough':'Borough', 'Neighborhood': 'Neighborhood'}, inplace=True)
df.head()

PostalCode           Borough     Neighbourhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

As the first step, I delete cells if their Borough is 'Not assigned'.

In [0]:
df_assigned = df[df['Borough'] != 'Not assigned']
df_assigned.head()

PostalCode           Borough     Neighbourhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M5A  Downtown Toronto       Regent Park
6        M6A        North York  Lawrence Heights

In [0]:
df_assigned.shape

(211, 3)

Now, the data is down to 211 rows. The first stage deleted 77 rows where Borough is 'Not assigned.' 

Second step: More than one neighborhood can exist in one postal code area. Such rows will be combined into one row with the neighborhoods separated with a comma. 

First, I determine how many unique postal codes exist in the dataframe. 

In [0]:
len(df_assigned['PostalCode'].unique())

103

There are 103 unique postal codes. That is, once I group rows with the same postal code in the same row, my dataframe should also have 103 rows. 

In [0]:
df_postal = df_assigned.groupby(['PostalCode', 'Borough'])['Neighbourhood'].apply(', '.join).reset_index()
df_postal.head()

PostalCode      Borough                           Neighbourhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [0]:
df_postal.shape

(103, 3)

Finally, if a cell has a borough but a Not assigned neighborhood, then I change the neighborhood to the borough. 

I first check if there are any such neighborhoods. 

In [0]:
df_postal[df_postal['Neighbourhood'] == 'Not assigned']

PostalCode       Borough Neighbourhood
85        M7A  Queen's Park  Not assigned

Seeing that there is one such neighborhood, I create a variable that will hold the borough information, and then change the value for neighborhood to the borough. 

In [0]:
boro = df_postal[df_postal['Neighbourhood'] == 'Not assigned']['Borough']
boro

85    Queen's Park
Name: Borough, dtype: object

In [0]:
df_postal.loc[df_postal['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = boro

As the last step, in the last cell of this part of the notebook, I find the shape of the dataframe. 

In [0]:
df_postal.shape

(103, 3)

## Part Two - Latitudes and Longitudes of Toronto Neighborhoods

First, I will download the coordinates for the postal codes in the dataframe.

In [0]:
coords = pd.read_csv('Geospatial.csv')
coords

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

I check the shape of the data frame to make sure that it has the same shape. I also note that both df_postal and coords are sorted according to their PostalCode column. 

In [0]:
coords.shape

(103, 3)

Then, I add the latitude and longitude from this coords to df_postal. 

In [0]:
df_postal['Latitude'] = coords['Latitude']
df_postal['Longitude'] = coords['Longitude']
df_postal.head()

PostalCode      Borough                           Neighbourhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [0]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_postal['Borough'].unique()),
        df_postal.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


Here is the resulting data frame.

In [0]:
df_postal

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    G

## Part Three - Clustering Neighborhoods

In this last part of the project, I will closely follow the steps in the Lab using the data on the New York City neighborhoods and venues making necessary changes for the Toronto data. 

First, I use geopy library to get the latitude and longitude values of Toronto.

In [0]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


Now, I create a map of Toronto using the latitude and longitude values from above. 

In [0]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_postal['Latitude'], df_postal['Longitude'], df_postal['Borough'], df_postal['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

As suggested in the guidelines, I will simplify the map by focusing on the neighborhoods that contain the word Toronto.

First, I will slice the original dataframe and create a new dataframe of the Toronto neighborhoods.

In [0]:
toronto = df_postal[df_postal['Borough'].str.contains('Toronto')]
toronto.head()

PostalCode          Borough                   Neighbourhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

In [0]:
toronto.shape

(38, 5)

The resulting dataframe has 38 postal codes. 

Next, I will create a map only for the postal code in this smaller dataframe, toronto. 

In [0]:
# create map of Toronto using latitude and longitude values
map_trt = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto['Latitude'], toronto['Longitude'], toronto['Borough'], toronto['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_trt)  
    
map_trt

Next, I will use the Foursquare API to explore the neighborhoods and segment them.

I define Foursquare Credentials and Version in the hidden cell below. 

In [0]:
CLIENT_ID = 'D3QIHY0IXET4VCGE4WL1HBBBHT1ZPNNSNKXTPSM2YNQ1M3BO' 
CLIENT_SECRET = 'PXZ0R0JR52GDDY5THHHLCWWOI2JWLQBZEZBNWQLWVXVDSXBQ' 
VERSION = '20180604'

Following the lab example, I explore venues within a radius of 500 meters and I will limit the number of venues per neighborhood to 100.

In [0]:
LIMIT = 100
radius = 500

Again, following the example, I create a function that will make a get request, will cleand the json data, and put it in a pandas dataframe for all our neighborhoods. 

In [0]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Then, I run this function on each neighborhood and create a new dataframe called toronto_venues.

In [0]:
toronto_venues = getNearbyVenues(names=toronto['Neighbourhood'],
                                   latitudes=toronto['Latitude'],
                                   longitudes=toronto['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

I check the shpae of the dataframe. There are 1708 venues extracted. 

In [0]:
print(toronto_venues.shape)
toronto_venues.head()

(1708, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                 Glen Stewart Ravine       43.676300       -79.294784   
4                       Upper Beaches       43.680563       -79.292869   

         Venue Category  
0                 Trail  
1     Health Food Store  
2                   Pub  
3  Other Great Outdoors  
4          Neighborhood

In [0]:
#toronto_venues.groupby('Neighborhood').count()

In [0]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 239 unique categories.


Now, I create dummies for each of these categories. 

In [0]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Beer Bar  Beer Store  Belgian Restaurant  Bistro  Board Shop  \
0         0           0                   0       0           0   
1         0           0                   0       0           0   
2         0           0                   0       0           0   
3         0           0                   0       0           0   
4         0           0                   0       0           0   

   Boat or Ferry  Bookstore  Boutique  Bowling Alley  Brazilian Restaurant  \
0              0          0         0              0                     0   
1              0          0         0              0                     0   
2              0          0         0              0                     0   
3              0          0         0              0                     0   
4              0          0         0              0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Bus Stop  Butcher  Café  \
0              0         0         0        0     0   
1              0         0         0        0     0   
2              0         0         0        0     0   
3              0         0         0        0     0   
4              0         0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  Cheese Shop  \
0

In [0]:
toronto_onehot.shape

(1708, 239)

In [0]:
# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

In [0]:
# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

Next, I group rows by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [0]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Wine Bar  Wings Joint  \
0                           Adelaide, King, Richmond      0.01          0.0   
1                                        Berczy Park      0.00          0.0   
2       Brockton, Exhibition Place, Parkdale Village      0.00          0.0   
3  Business Reply Mail Processing Centre 969 Eastern      0.00          0.0   
4  CN Tower, Bathurst Quay, Island airport, Harbo...      0.00          0.0   

   Yoga Studio  Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0     0.000000                0.0  0.000000            0.000000      0.000000   
1     0.000000                0.0  0.000000            0.000000      0.000000   
2     0.047619                0.0  0.000000            0.000000      0.000000   
3     0.000000                0.0  0.000000            0.000000      0.000000   
4     0.000000                0.0  0.058824            0.058824      0.058824   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.000000         0.000000          0.000000                 0.04   
1        0.000000         0.000000          0.000000                 0.00   
2        0.000000         0.000000          0.000000                 0.00   
3        0.000000         0.000000          0.000000                 0.00   
4        0.117647         0.176471          0.117647                 0.00   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0     0.010000        0.01                  0.0   
1           0.0       0.0     0.017241        0.00                  0.0   
2           0.0       0.0     0.000000        0.00                  0.0   
3           0.0       0.0     0.000000        0.00                  0.0   
4           0.0       0.0     0.000000        0.00                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0              0.03                 0.0       0.000000   0.000000         0.0   
1              0.00                 0.0       0.000000   0.017241         0.0   
2              0.00                 0.0       0.000000   0.000000         0.0   
3              0.00                 0.0       0.052632   0.000000         0.0   
4              0.00                 0.0       0.000000   0.000000         0.0   

   Bagel Shop    Bakery  Bank       Bar  Baseball Stadium  Basketball Stadium  \
0    0.000000  0.030000   0.0  0.040000               0.0            0.000000   
1    0.017241  0.034483   0.0  0.000000               0.0            0.017241   
2    0.000000  0.000000   0.0  0.047619               0.0            0.000000   
3    0.000000  0.000000   0.0  0.000000               0.0            0.000000   
4    0.000000  0.000000   0.0  0.058824               0.0            0.000000   

      Beach  Beer Bar  Beer Store  Belgian Restaurant    Bistro  Board Shop  \
0  0.000000  0.000000         0.0            0.000000  0.000000         0.0   
1  0.017241  0.034483         0.0            0.017241  0.017241         0.0   
2  0.000000  0.000000         0.0            0.000000  0.000000         0.0   
3  0.000000  0.000000         0.0            0.000000  0.000000         0.0   
4  0.000000  0.000000         0.0            0.000000  0.000000         0.0   

   Boat or Ferry  Bookstore  Boutique  Bowling Alley  Brazilian Restaurant  \
0       0.000000       0.01  0.000000            0.0                  0.01   
1       0.000000       0.00  0.000000            0.0                  0.00   
2       0.000000       0.00  0.000000            0.0                  0.00   
3       0.000000       0.00  0.000000            0.0                  0.00   
4       0.058824       0.00  0.058824            0.0                  0.00   

   Breakfast Spot   Brewery  Bubble Tea Shop  Building  Burger Joint  \
0        0.020000  0.000000              0.0       0.0          0.03   
1        0.017241  0.000000              0.0       0.0          0.00   
2        0.095238  0.000000   

I check to make sure the shape of the dataframe is as expected. 

In [0]:
toronto_grouped.shape

(38, 239)

Next, I use the function from the lab to sort the venues in descending order.

In [0]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Then, I create a new dataframe with the top 10 venues for each neighborhood.

In [0]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café   American Restaurant            Steakhouse   
1          Cocktail Bar    Seafood Restaurant                  Café   
2        Breakfast Spot           Coffee Shop  Caribbean Restaurant   
3  Gym / Fitness Center           Pizza Place            Smoke Shop   
4        Airport Lounge      Airport Terminal           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar       Thai Restaurant          Burger Joint   
1            Steakhouse           Cheese Shop        Farmers Market   
2             Pet Store          Intersection                   Bar   
3            Skate Park               Brewery            Restaurant   
4      Sculpture Garden                 Plane         Boat or Ferry   

  8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0                   Gym        Cosmetics Shop        Sushi Restaurant  
1                Bakery              Beer Bar          Breakfast Spot  
2         Burrito Place            Restaurant  Furniture / Home Store  
3      Recording Studio         Burrito Place              Comic Shop  
4                   Bar          Airport Gate      Airport Food Court

Finally, I cluster neighborhoods using k-means clustering. I try different k values, but they all produce almost the same outcome where most neighborhoods are grouped within the same cluster with a few different ones. 

Here I present the results with k=4. 

In [0]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

I create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [0]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

toronto_merged.head() 

PostalCode          Borough                   Neighbourhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37 -79.293031               0  Other Great Outdoors                   Pub   
41 -79.352188               2      Greek Restaurant           Coffee Shop   
42 -79.315572               2          Liquor Store               Brewery   
43 -79.340923               2                  Café           Coffee Shop   
44 -79.388790               1              Bus Line           Swim School   

   3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
37                 Trail     Health Food Store          Discount Store   
41    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
42        Ice Cream Shop            Steakhouse    Fast Food Restaurant   
43                Bakery             Gastropub      Italian Restaurant   
44                  Park    College Rec Center    Colombian Restaurant   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
37      Department Store                 Dessert Shop    Dim Sum Restaurant   
41               Brewery                          Pub                   Spa   
42    Italian Restaurant             Sushi Restaurant         Burrito Place   
43           Music Store          American Restaurant               Brewery   
44     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   9th Most Common Venue 10th Most Common Venue  
37                 Diner                Dog Run  
41     Indian Restaurant                 Bakery  
42          Burger Joint      Fish & Chips Shop  
43                  Bank                    Bar  
44            Donut Shop       Doner Restaurant

Finally, I use folium to depict the 4 clusters. 

In [0]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Cluster 1

In [0]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
37     East Toronto               0  Other Great Outdoors   
48  Central Toronto               0            Restaurant   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
37                   Pub                 Trail      Health Food Store   
48                 Trail            Playground  Vietnamese Restaurant   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
37               Discount Store      Department Store          Dessert Shop   
48  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
37    Dim Sum Restaurant                 Diner                Dog Run  
48      Doner Restaurant               Dog Run               Dive Bar

#### Cluster 2

In [0]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
44   Central Toronto               1              Bus Line   
50  Downtown Toronto               1                  Park   
64   Central Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
44           Swim School                  Park    College Rec Center   
50                 Trail            Playground              Building   
64                 Trail          Home Service         Jewelry Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
44  Colombian Restaurant     Electronics Store  Eastern European Restaurant   
50          Dance Studio   Dumpling Restaurant                   Donut Shop   
64      Sushi Restaurant            Donut Shop             Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
44   Dumpling Restaurant            Donut Shop       Doner Restaurant  
50      Doner Restaurant               Dog Run               Dive Bar  
64               Dog Run   Dumpling Restaurant           Dance Studio

#### Cluster 3

In [0]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
41      East Toronto               2      Greek Restaurant   
42      East Toronto               2          Liquor Store   
43      East Toronto               2                  Café   
45   Central Toronto               2                   Gym   
46   Central Toronto               2           Coffee Shop   
47   Central Toronto               2        Sandwich Place   
49   Central Toronto               2           Coffee Shop   
51  Downtown Toronto               2           Coffee Shop   
52  Downtown Toronto               2           Coffee Shop   
53  Downtown Toronto               2           Coffee Shop   
54  Downtown Toronto               2           Coffee Shop   
55  Downtown Toronto               2           Coffee Shop   
56  Downtown Toronto               2           Coffee Shop   
57  Downtown Toronto               2           Coffee Shop   
58  Downtown Toronto               2           Coffee Shop   
59  Downtown Toronto               2           Coffee Shop   
60  Downtown Toronto               2           Coffee Shop   
61  Downtown Toronto               2           Coffee Shop   
65   Central Toronto               2           Coffee Shop   
66  Downtown Toronto               2                  Café   
67  Downtown Toronto               2                  Café   
68  Downtown Toronto               2       Airport Service   
69  Downtown Toronto               2           Coffee Shop   
70  Downtown Toronto               2           Coffee Shop   
75  Downtown Toronto               2         Grocery Store   
76      West Toronto               2                Bakery   
77      West Toronto               2                   Bar   
78      West Toronto               2                  Café   
82      West Toronto               2    Mexican Restaurant   
83      West Toronto               2             Gift Shop   
84      West Toronto               2           Coffee Shop   
87      East Toronto               2    Light Rail Station   

            2nd Most Common Venue  3rd Most Common Venue  \
41                    Coffee Shop     Italian Restaurant   
42                        Brewery         Ice Cream Shop   
43                    Coffee Shop                 Bakery   
45                          Hotel         Clothing Store   
46           Gym / Fitness Center    Sporting Goods Shop   
47                   Dessert Shop            Pizza Place   
49                            Pub    Fried Chicken Joint   
51                    Pizza Place                   Park   
52            Japanese Restaurant                Gay Bar   
53                           Park                 Bakery   
54                 Clothing Store         Cosmetics Shop   
55                          Hotel             Restaurant   
56                   Cocktail Bar     Seafood Restaurant   
57             Italian Restaurant                   Café   
58                           Café    American Restaurant   
59                          Hotel               Aquarium   
60                           Café                  Hotel   
61                           Café                  Hotel   
65                 Sandwich Place                   Café   
66                      Bookstore                    Bar   
67  Vegetarian / Vegan Restaurant  Vietnamese Restaurant   
68                 Airport Lounge       Airport Terminal   
69                           Café             Restaurant   
70                           Café             Steakhouse   
75                           Café                   Park   
76                       Pharmacy            Supermarket   
77                    Coffee Shop       Asian Restaurant   
78                 Breakfast Spot            Coffee Shop   
82                            Bar                   Café   
83                 Breakfast Spot            Coffee Shop   
84                    Pizza Place                   Café   
87           Gym / Fitness Center            P

#### Cluster 4

In [0]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
63  Central Toronto               3                  Pool   

   2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
63                Garden  Vietnamese Restaurant          Dance Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
63  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
63      Doner Restaurant               Dog Run               Dive Bar

### Conclusion

Based on the clustering results, it is possible to argue that Toronto neighborhoods are similar to each other in terms of the venues available in each neighborhood. Out of the 38 neighborhoods, only 6 belongs to different clusters when the number of clusters is set to 4. The numbers are not different even when the number of clusters is 10. 

**Thanks for reading!**